core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java

```diff
@@ -101,29 +107,161 @@
   private final boolean inHeap;
 
   /**
+   * The size of memory granted to each consumer.
+   */
+  private final HashMap<MemoryConsumer, Long> consumers;
+
+  /**
    * Construct a new TaskMemoryManager.
    */
   public TaskMemoryManager(MemoryManager memoryManager, long taskAttemptId) {
     this.inHeap = memoryManager.tungstenMemoryIsAllocatedInHeap();
     this.memoryManager = memoryManager;
     this.taskAttemptId = taskAttemptId;
+    this.consumers = new HashMap<>();
   }
 
   /**
-   * Acquire N bytes of memory for execution, evicting cached blocks if necessary.
+   * Acquire N bytes of memory for a consumer. If there is no enough memory, it will call
+   * spill() of consumers to release more memory.
+   *
    * @return number of bytes successfully granted (<= N).
    */
-  public long acquireExecutionMemory(long size) {
-    return memoryManager.acquireExecutionMemory(size, taskAttemptId);
+  public long acquireExecutionMemory(long required, MemoryConsumer consumer) {
+    assert(required >= 0);
+    synchronized (this) {
+      long got = memoryManager.acquireExecutionMemory(required, taskAttemptId);
+
+      // try to release memory from other consumers first, then we can reduce the frequency of
+      // spilling, avoid to have too many spilled files.
+      if (got < required) {
+        // consumers could be modified by spill(), so we should have a copy here.
+        MemoryConsumer[] cs = new MemoryConsumer[consumers.size()];
+        consumers.keySet().toArray(cs);
+        // Call spill() on other consumers to release memory
+        for (MemoryConsumer c: cs) {
+          if (c != null && c != consumer) {
```

Comment:
There could be some cases the memory is acquire not by a consumer, it will be null.